### PR TITLE
Connection handling with shutdown

### PR DESCRIPTION
**Pull request by replay**
Adding a method to shutdown the inputs cleanly before exiting

### CODE REVIEW

**Comment by replay** [comment]
That's not tested yet, except the unit tests that I added, but they are not complete. 

If you get a chance, please take a look @DanCech. Once you approve that it looks mergable I'll do a bunch of manual tests.

**Comment by DanCech** [comment]
I'm going to have to spend some time wrapping my head around this

**Comment by Dieterbe** [comment]
out of scope for this PR, but I think the input plugins should separate their "start" logic from the construction of the instance.

File: input/workerpool.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	wg       sync.WaitGroup
+	shutdown chan struct{}
+}
+
+// registers this instance in the list of stoppables
+// also initializes the shutdown channel
+func (w *workerPool) startStoppable() {
+	stoppables = append(stoppables, w)
+	w.shutdown = make(chan struct{})
+}
```
Comment (Dieterbe) [inline input/workerpool.go]: this is hard to follow. instance means "a pool of workers", which adds itself to a list which is defined in a different file?
and "startstoppable" is not a verb.

Comment (Dieterbe) [reply input/workerpool.go]: this is too abstract.
I think the caller (e.g. the main command) should maintain a list of plugins. and just add them to the list when it creates them.
and the input plugins can have their own stop methods and waitgroups if they need them, instead of embedding this vague separate type. and the instances in the input package should also not have their own determination of when they timeout, that can be handled by the caller, as pointed out in another comment.

this is basically how MT does it, and i think it's easier to follow

File: input/listen.go [hunk 2]
```diff
@@ -8,126 +8,161 @@ import (
...
+				err := reconnect(addr)
+				if err == nil {
+					backoffCounter.Reset()
+					break
+				}
+
+				backoffDuration := backoffCounter.Duration()
+				log.Error("error listening on %v/%s, retrying after %v: %s", addr, proto, backoffDuration, err)
+				select {
+				case <-l.shutdown:
```
Comment (Dieterbe) [inline input/listen.go]: shouldn't we check this case before we log the above backoffDuration call & error/retry message?

File: input/workerpool.go (hunk 1 above)
Comment (replay) [reply input/workerpool.go]: just as an explanation of the name:
`Stoppable` is the type of the object that have a `.stop()` method, this method starts the `Stoppable`, so `startStoppable()`.

File: input/listen.go [hunk 3]
```diff
@@ -8,126 +8,161 @@ import (
...
-			l, err = listenTcp(addr)
-			if err == nil {
-				backoffCounter.Reset()
-				break
+func (l *Listener) acceptTcp(addr string) {
+	for {
+		c, err := l.tcpList.AcceptTCP()
+		if err != nil {
+			select {
+			// socket has been closed due to shut down, log and return
```
Comment (Dieterbe) [inline input/listen.go]: this comment is incorrect. and if it were correct it would be redundent, so just remove it.

File: input/listen.go [hunk 4]
```diff
@@ -8,126 +8,161 @@ import (
...
-
-		for {
-			// read a packet into buffer
-			b, src, err := l.ReadFrom(buffer)
-			if err != nil {
+		// read a packet into buffer
+		b, src, err := l.udpConn.ReadFrom(buffer)
+		if err != nil {
+			select {
+			// socket has been closed due to shut down, log and return
```
Comment (Dieterbe) [inline input/listen.go]: this comment is incorrect. and if it were correct it would be redundent, so just remove it.

File: input/listen.go [hunk 5]
```diff
@@ -8,126 +8,161 @@ import (
...
+	for {
+		log.Notice("listening on %v/%s", addr, proto)
+
+		consume(addr)
+
+		select {
+		case <-l.shutdown:
+			return
+		default:
+			for {
```
Comment (Dieterbe) [inline input/listen.go]: can save a level of indent here by making the default case empty in the select (moving the for loop after the select)

File: input/listen.go [hunk 6]
```diff
@@ -8,126 +8,161 @@ import (
...
 }
 
-func listenTcp(addr string) (*net.TCPListener, error) {
-	laddr, err := net.ResolveTCPAddr("tcp", addr)
-	if err != nil {
-		return nil, err
-	}
-	return net.ListenTCP("tcp", laddr)
-}
+func (l *Listener) accept(addr, proto string, consume func(string), reconnect func(string) error, listener Closable) {
```
Comment (Dieterbe) [inline input/listen.go]: why is this function called accept? something like `run` seems more appropriate and less confusing.

File: input/amqp.go [hunk 7]
```diff
@@ -104,22 +96,71 @@ func connectAMQP(a *Amqp) (<-chan amqp.Delivery, error) {
...
+	defer a.wg.Done()
+
+	for {
+		c, err := a.connect(a)
+		if err != nil {
+			// failed to connect; backoff and try again
+			log.Error("connectAMQP: %v", err)
+
+			d := b.Duration()
+			log.Info("retrying in %v", d)
```
Comment (Dieterbe) [inline input/amqp.go]: message should go after shutdown check

File: cmd/carbon-relay-ng/carbon-relay-ng.go [hunk 8]
```diff
@@ -221,6 +221,11 @@ func main() {
 	}
 
 	select {}
+
```
Comment (Dieterbe) [inline cmd/carbon-relay-ng/carbon-relay-ng.go]: how is this supposed to work? i don't think the code will advance past line 223

File: input/amqp_test.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+	}
+}
+
+func TestAmqpFailingShutdown(t *testing.T) {
+	dispatcher := mockDispatcher{}
+	c, mockConn, mockChan, mockConnector := getMockConnector()
+	a := NewAMQP(config, &dispatcher, mockConnector)
+	go a.Start()
+
+	dispatcher.dispatchDuration = time.Second * 5
```
Comment (Dieterbe) [inline input/amqp_test.go]: this property isn't used anywhere

File: input/amqp_test.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+	os.Exit(res)
+}
+
+func TestAmqpSuccessfulShutdown(t *testing.T) {
+	dispatcher := mockDispatcher{}
+	c, mockConn, mockChan, mockConnector := getMockConnector()
+	a := NewAMQP(config, &dispatcher, mockConnector)
+	go a.Start()
+
+	dispatcher.dispatchDuration = time.Millisecond
```
Comment (Dieterbe) [inline input/amqp_test.go]: this property isn't used anywhere

**Comment by Dieterbe** [comment]
with @replay on vacation i can probably resume/finish this work. not quite sure how the dispatchduration/amqp testing is supposed to work but maybe i can figure it out if i look at it again.

File: cmd/carbon-relay-ng/carbon-relay-ng.go (hunk 8 above)
Comment (DanCech) [reply cmd/carbon-relay-ng/carbon-relay-ng.go]: Not sure I understand this comment

Comment (Dieterbe) [reply cmd/carbon-relay-ng/carbon-relay-ng.go]: line 223 was `select {}` which blocks indefinitely and thus prevented the new code that @replay added from running. i've changed this to use a signal handler and trigger shutdown on certain signals

**Comment by Dieterbe** [comment]
@DanCech this should be good to merge now
i did some manual testing of the tcp happy path as well.
